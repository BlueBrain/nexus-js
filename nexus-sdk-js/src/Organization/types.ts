export interface ListOrgOptions {
  full_text_search?: string;
  from?: number;
  size?: number;
  deprecated?: boolean;
  [key: string]: any;
}

export interface CreatOrgPayload {
  description?: string;
}
export type Context = string | string[];
