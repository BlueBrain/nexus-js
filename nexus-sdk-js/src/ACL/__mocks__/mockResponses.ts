import { ACLResponse, ListACLResponse } from '../types';

export const mockACLResponse: ACLResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/resource.json',
    'https://bluebrain.github.io/nexus/contexts/acls.json',
  ],
  '@id': 'https://nexus.example.com/v1/acls/org1',
  '@type': 'User',
  _path: '/org1',
  acl: [
    {
      permissions: ['read'],
      identity: {
        '@id': 'https://nexus.example.com/v1/realms/myrealm/groups/a-group',
        '@type': 'Group',
        realm: 'myrealm',
        group: 'a-group',
      },
    },
    {
      permissions: ['read', 'write'],
      identity: {
        '@id': 'https://nexus.example.com/v1/realms/myrealm/groups/some-group',
        '@type': 'Group',
        realm: 'myrealm',
        group: 'some-group',
      },
    },
    {
      permissions: ['acls/read', 'acls/write'],
      identity: {
        '@id': 'https://nexus.example.com/v1/realms/myrealm/users/user',
        '@type': 'User',
        realm: 'myrealm',
        subject: 'alice',
      },
    },
  ],
  _createdAt: '2018-09-18T09:58:00.801Z',
  _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _updatedAt: '2018-09-18T10:01:00.801Z',
  _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _rev: 1,
};

export const mockListACLResponse: ListACLResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/resource.json',
    'https://bluebrain.github.io/nexus/contexts/iam.json',
    'https://bluebrain.github.io/nexus/contexts/search.json',
  ],
  _total: 2,
  _results: [
    {
      '@id': 'https://nexus.example.com/v1/acls/myorg/myproj',
      '@type': 'Authenticated',
      acl: [
        {
          permissions: ['read', 'write'],
          identity: {
            '@id': 'https://nexus.example.com/v1/realm/groups/two',
            '@type': 'Group',
            realm: 'myrealm',
            group: 'two',
          },
        },
      ],
      _path: '/myorg/myproj',
      _rev: 1,
      _createdAt: '2018-09-17T14:55:42.939Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
      _updatedAt: '2018-09-17T15:05:42.939Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
    },
    {
      '@id': 'https://nexus.example.com/v1/acls/myorg/myproj2',
      '@type': 'Anonymous',
      acl: [
        {
          permissions: ['read'],
          identity: {
            '@id': 'https://nexus.example.com/v1/realms/myrealm/users/me',
            '@type': 'User',
            realm: 'myrealm',
            subject: 'me',
          },
        },
      ],
      _path: '/myorg/myproj2',
      _rev: 2,
      _createdAt: '2018-09-17T14:00:42.939Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/alice',
      _updatedAt: '2018-09-17T14:05:42.939Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/alice',
    },
    {
      '@id': 'https://nexus.example.com/v1/acls/myorg/myproj',
      '@type': 'Authenticated',
      acl: [
        {
          permissions: ['read', 'write'],
          identity: {
            '@id': 'https://nexus.example.com/v1/realm/groups/two',
            '@type': 'Group',
            realm: 'myrealm',
            group: 'two',
          },
        },
      ],
      _path: '/myorg/myproj',
      _rev: 1,
      _createdAt: '2018-09-17T14:55:42.939Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
      _updatedAt: '2018-09-17T15:05:42.939Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
    },
    {
      '@id': 'https://nexus.example.com/v1/acls/myorg/myproj2',
      '@type': 'Anonymous',
      acl: [
        {
          permissions: ['read'],
          identity: {
            '@id': 'https://nexus.example.com/v1/realms/myrealm/users/me',
            '@type': 'User',
            realm: 'myrealm',
            subject: 'me',
          },
        },
      ],
      _path: '/myorg/myproj2',
      _rev: 2,
      _createdAt: '2018-09-17T14:00:42.939Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/alice',
      _updatedAt: '2018-09-17T14:05:42.939Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/alice',
    },
  ],
};
