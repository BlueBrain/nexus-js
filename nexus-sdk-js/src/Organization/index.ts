import Project from '../Project';
import {
  createOrganization,
  getOrganization,
  listOrganizations,
  deprecateOrganization,
  updateOrganization,
} from './utils';
import { Context } from './types';

export interface OrgResponseCommon {
  '@id': string;
  '@type': string;
  _uuid: string;
  _label: string;
  _rev: number;
  _deprecated: boolean;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _self?: string;
  _constrainedBy?: string;
  description?: string;
}

export interface ListOrgResponse {
  '@context': Context;
  _total: number;
  _results?: OrgResponseCommon[];
  _links?: any;
  code?: string;
  message?: string;
}

export interface OrgResponse extends OrgResponseCommon {
  '@context': Context;
}

export default class Organization {
  context?: Context;
  id: string;
  type: string;
  uuid: string;
  label: string;
  rev: number;
  deprecated: boolean;
  createdAt: string;
  createdBy: string;
  updatedAt: string;
  updatedBy: string;
  self?: string;
  constrainedBy?: string;
  description?: string;

  static get = getOrganization;
  static list = listOrganizations;
  static create = createOrganization;
  static update = updateOrganization;
  static deprecate = deprecateOrganization;

  constructor(organizationResponse: OrgResponse) {
    this.context = organizationResponse['@context'];
    this.id = organizationResponse['@id'];
    this.type = organizationResponse['@type'];
    this.uuid = organizationResponse._uuid;
    this.label = organizationResponse._label;
    this.rev = organizationResponse._rev;
    this.deprecated = organizationResponse._deprecated;
    this.createdAt = organizationResponse._createdAt;
    this.createdBy = organizationResponse._createdBy;
    this.updatedAt = organizationResponse._updatedAt;
    this.updatedBy = organizationResponse._updatedBy;
    this.self = organizationResponse._self;
    this.constrainedBy = organizationResponse._constrainedBy;
    this.description = organizationResponse.description;
  }

  async getProject(projectLabel: string): Promise<Project> {
    return Project.get(this.label, projectLabel);
  }

  async listProjects(): Promise<Project[]> {
    return Project.list(this.label);
  }
}
