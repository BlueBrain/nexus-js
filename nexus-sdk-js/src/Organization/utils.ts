import { Resource, Organization } from '..';
import { OrgResponse, ListOrgsResponse, ListOrgsOptions } from '.';
import { httpPut, httpGet, httpDelete } from '../utils/http';
import { CreateOrganizationException } from './exceptions';

/**
 *
 * @param label The label of your organization
 * @param name The name of your organization
 */
export async function createOrganization(
  label: string,
  name: string,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(`/orgs/${label}`, {
      name,
    });
    return new Organization({
      ...orgResponse,
      projectNumber: 0,
      _deprecated: false,
    });
  } catch (error) {
    throw new CreateOrganizationException(error.message);
  }
}

// TODO: refactor -> blocked by https://github.com/BlueBrain/nexus/issues/112
/**
 *
 * @param orgLabel The organization label to fetch
 * @param Object<revision, tag> The specific tag OR revision to fetch
 */
export async function getOrganization(
  orgLabel: string,
  options?: { revision?: number; tag?: string },
): Promise<Organization> {
  try {
    // check if we have options
    let ops = '';
    if (options) {
      // it's rev or tag, not both. We take rev over tag
      if (options.revision) {
        ops = `?rev=${options.revision}`;
      } else if (options.tag) {
        ops = `?tag=${options.tag}`;
      }
    }
    const orgResponse: OrgResponse = await httpGet(`/orgs/${orgLabel}${ops}`);
    // we want to know how many projects there are per organisation
    const listOrgsResponse: ListOrgsResponse = await httpGet('/projects');
    const projectNumber: number = listOrgsResponse._results
      ? listOrgsResponse._results.reduce((prev, org) => {
          const split = org._id.split('/');
          const orgName = split.slice(split.length - 2, split.length - 1)[0];
          if (orgName === orgLabel) {
            return prev + 1;
          }
          return prev;
        }, 0)
      : 0;

    const org = new Organization({ ...orgResponse, projectNumber });
    return org;
  } catch (e) {
    throw new Error(`ListOrgsError: ${e}`);
  }
}

// TODO: refactor -> blocked by https://github.com/BlueBrain/nexus/issues/112
// cannot implement list orgs options until then...
export async function listOrganizations(
  options?: ListOrgsOptions,
): Promise<Organization[]> {
  let ops;
  if (options) {
    ops = Object.keys(options).reduce(
      (currentOps, key) =>
        currentOps.length === 0
          ? `?${key}=${options[key]}`
          : `${currentOps}&${key}=${options[key]}`,
      '',
    );
  }
  try {
    const listOrgsResponse: ListOrgsResponse = await httpGet('/projects');
    if (listOrgsResponse.code || !listOrgsResponse._results) {
      return [];
    }
    // Get list of unique orgs names
    const filteredOrgNames: string[] = listOrgsResponse._results
      .map(org => {
        const split = org._id.split('/');
        const orgName = split.slice(split.length - 2, split.length - 1)[0];
        return orgName;
      })
      .filter((org, index, self) => self.indexOf(org) === index);

    // get orgs details
    return Promise.all(
      filteredOrgNames.map(async org => await getOrganization(org)),
    );
  } catch (e) {
    throw new Error(`ListOrgsError: ${e}`);
  }
}

/**
 *
 * @param orgLabel Current organization label
 * @param rev Last known revision
 * @param newName Knew name the organization will be called with
 */
export async function updateOrganization(
  orgLabel: string,
  rev: number = 1,
  newName: string,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(
      `/orgs/${orgLabel}?rev=${rev}`,
      {
        name: newName,
      },
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 *
 * @param orgLabel Current organization label
 * @param rev Last know revision
 * @param Object<{tagName, revision}> The name of the tag and revision number to tag the organization from
 */
export async function tagOrganization(
  orgLabel: string,
  rev: number = 1,
  {
    tagName,
    tagFromRev,
  }: {
    tagName: string;
    tagFromRev: number;
  },
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(
      `/orgs/${orgLabel}/tags?rev=${rev}`,
      {
        tag: tagName,
        rev: tagFromRev,
      },
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 *
 * @param orgLabel Label of the Org to be deprecated
 * @param rev Last know revision
 */
export async function deprecateOrganization(
  orgLabel: string,
  rev: number = 1,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpDelete(
      `/orgs/${orgLabel}?rev=${rev}`,
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}
