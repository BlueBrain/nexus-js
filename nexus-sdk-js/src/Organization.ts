import { httpGet } from './utils/http';
import Project, { ProjectResponse } from './Project';

export interface ListOrgsResponse {
  '@context': string;
  _total: number;
  _links?: string;
  _results?: [{ _id: string }];
  code?: string;
  message?: string;
}

export interface OrgInit {
  '@context': string;
  '@id': string;
  _label: string;
}

export default class Organization {
  context: string;
  id: string;
  label: string;

  constructor(organizationResponse: OrgInit) {
    this.context = organizationResponse['@context'];
    this.id = organizationResponse['@id'];
    this.label = organizationResponse._label;
  }

  async listProjects(): Promise<Project[]> {
    try {
      const projects = await httpGet(`/projects/${this.label}`);
      return projects.map((project: ProjectResponse) => new Project(project));
    } catch (e) {
      return e;
    }
  }
}
