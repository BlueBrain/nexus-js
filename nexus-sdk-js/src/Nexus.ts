import Store from './utils/Store';
import Organization, { OrganizationResponse } from './Organization';
import { httpGet, httpPost } from './utils/http';

type NexusConfig = {
  environment: string;
  token?: string;
};

export const store: Store = new Store({
  auth: {
    accessToken: null,
  },
  api: {
    baseUrl: null,
  },
});

export default class Nexus {
  constructor(config: NexusConfig) {
    if (!config.environment) {
      throw new Error(
        'No environment provided. Please specify your Nexus instance endpoint.'
      );
    }
    store.update('api', state => ({
      ...state,
      baseUrl: config.environment,
    }));
    if (config.token) {
      store.update('auth', state => ({
        ...state,
        accessToken: config.token,
      }));
    }
  }

  setToken(token: string): void {
    store.update('auth', state => ({
      ...state,
      accessToken: token,
    }));
  }

  removeToken(): void {
    store.update('auth', state => ({
      ...state,
      accessToken: undefined,
    }));
  }

  async listOrganizations(): Promise<Organization[]> {
    try {
      const orgs: OrganizationResponse[] = await httpGet('/orgs');
      return orgs.map((org: OrganizationResponse) => new Organization(org));
    } catch (e) {
      console.log(e);
      return e;
    }
  }

  async getOrganization(id: string): Promise<Organization> {
    try {
      const org: OrganizationResponse = await httpGet(`/orgs/${id}`);
      return new Organization(org);
    } catch (e) {
      return e;
    }
  }

  async createOrganization(): Promise<Organization> {
    try {
      const org: OrganizationResponse = await httpPost('/orgs', {});
      return new Organization(org);
    } catch (e) {
      return e;
    }
  }
}
