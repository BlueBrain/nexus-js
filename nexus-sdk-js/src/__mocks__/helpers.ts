import { ProjectResponse, ListProjectsResponse } from '../Project';
import { ListOrgResponse, OrgResponse } from '../Organization';
import { ResourceResponse, ListResourceResponse } from '../Resource';
import { ViewsListResponse } from '../View/utils';
import {
  ElasticSearchViewResponse,
  ElasticSearchViewQueryResponse,
  ElasticSearchViewAggregationResponse,
} from '../View/ElasticSearchView';
import { SparqlViewResponse } from '../View/SparqlView';

export const mockOrgResponse: OrgResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/resource.json',
    'https://bluebrain.github.io/nexus/contexts/admin.json',
  ],
  '@id': 'https://nexus.example.com/v1/orgs/myorg',
  '@type': 'Organization',
  description: 'My Organization',
  _label: 'myorg',
  _uuid: 'bc0eba71-2a7f-40e8-ac90-5c97fc6f37d7',
  _rev: 4,
  _deprecated: true,
  _createdAt: '2018-09-18T09:58:00.801Z',
  _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _updatedAt: '2018-09-18T10:01:00.801Z',
  _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
};

export const mockListOrgResponse: ListOrgResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/admin.json',
    'https://bluebrain.github.io/nexus/contexts/search.json',
    'https://bluebrain.github.io/nexus/contexts/resource.json',
  ],
  _total: 2,
  _results: [
    {
      '@id': 'https://nexus.example.com/v1/orgs/myorg',
      '@type': 'Organization',
      description: 'My Organization',
      _label: 'myorg',
      _uuid: 'bc0eba71-2a7f-40e8-ac90-5c97fc6f37d7',
      _rev: 4,
      _deprecated: true,
      _createdAt: '2018-09-18T09:58:00.801Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
      _updatedAt: '2018-09-18T10:01:00.801Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
    },
    {
      '@id': 'https://nexus.example.com/v1/orgs/myorg2',
      '@type': 'Organization',
      description: 'My Second Organization',
      _label: 'myorg2',
      _uuid: 'b42e5206-f81e-430b-808d-5dac2599153d',
      _rev: 1,
      _deprecated: false,
      _createdAt: '2019-01-14T09:29:39.416Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
      _updatedAt: '2019-01-14T09:29:39.416Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
    },
  ],
};

export const mockProjectResponse: ProjectResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/resource.json',
    'https://bluebrain.github.io/nexus/contexts/admin.json',
  ],
  '@id': 'https://nexus.example.com/v1/projects/myorg/myproject',
  '@type': 'Project',
  base: 'https://nexus.example.com/v1/myorg/myproject/',
  vocab: 'https://schema.org/',
  apiMappings: [
    {
      prefix: 'person',
      namespace: 'http://example.com/some/person',
    },
    {
      prefix: 'schemas',
      namespace: 'https://bluebrain.github.io/nexus/schemas/',
    },
    {
      prefix: 'ex',
      namespace: 'http://example.com/',
    },
  ],
  _label: 'myproject',
  _organizationLabel: 'myorg',
  _organizationUuid: 'bc0eba71-2a7f-40e8-ac90-5c97fc6f37d7',
  _uuid: 'e622745d-5eea-4dc5-8318-58a3dd4101ff',
  _rev: 4,
  _deprecated: true,
  _createdAt: '2018-09-18T09:58:00.801Z',
  _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _updatedAt: '2018-09-18T10:30:00.801Z',
  _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
};

export const mockListProjectResponse: ListProjectsResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/admin.json',
    'https://bluebrain.github.io/nexus/contexts/search.json',
    'https://bluebrain.github.io/nexus/contexts/resource.json',
  ],
  _total: 1,
  _results: [
    {
      '@id': 'https://nexus.example.com/v1/projects/myorg/myproject',
      '@type': 'Project',
      base: 'http://nexus.example.com/',
      vocab: 'http://schema.org/',
      apiMappings: [
        {
          namespace: 'http://schema.org/',
          prefix: 'schema',
        },
      ],
      _uuid: 'e622745d-5eea-4dc5-8318-58a3dd4101ff',
      _label: 'myproject',
      _organizationUuid: 'bc0eba71-2a7f-40e8-ac90-5c97fc6f37d7',
      _organizationLabel: 'myorg',
      _rev: 4,
      _deprecated: true,
      _createdAt: '2018-09-18T09:58:00.801Z',
      _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
      _updatedAt: '2018-09-18T10:30:00.801Z',
      _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
    },
  ],
};

export const mockResourceResponse: ResourceResponse = {
  '@context': '',
  '@id': 'https://incf.github.io/neuroshapes/contexts/schema.json',
  _self:
    'https://bbp.epfl.ch/nexus/v1/resources/anorg/testcore/https%3A%2F%2Fbluebrain.github.io%2Fnexus%2Fschemas%2Fresource.json/context:schema.json',
  _constrainedBy: 'https://bluebrain.github.io/nexus/schemas/resource.json',
  _project: 'https://bbp.epfl.ch/nexus/v1/projects/anorg/testcore',
  _createdAt: '2018-11-14T21:16:54.230Z',
  _createdBy:
    'https://bbp.epfl.ch/nexus/v1/realms/BBP/users/f:9d46ddd6-134e-44d6-aa74-bdf00f48dfce:sy',
  _updatedAt: '2018-11-15T08:40:52.735Z',
  _updatedBy:
    'https://bbp.epfl.ch/nexus/v1/realms/BBP/users/f:9d46ddd6-134e-44d6-aa74-bdf00f48dfce:sy',
  _rev: 2,
  _deprecated: false,
};

export const mockListResourceResponse: ListResourceResponse = {
  '@context': '',
  _results: [
    {
      '@id': 'https://incf.github.io/neuroshapes/contexts/schema.json',
      _self:
        'https://bbp.epfl.ch/nexus/v1/resources/anorg/testcore/https%3A%2F%2Fbluebrain.github.io%2Fnexus%2Fschemas%2Fresource.json/context:schema.json',
      _constrainedBy: 'https://bluebrain.github.io/nexus/schemas/resource.json',
      _project: 'https://bbp.epfl.ch/nexus/v1/projects/anorg/testcore',
      _createdAt: '2018-11-14T21:16:54.230Z',
      _createdBy:
        'https://bbp.epfl.ch/nexus/v1/realms/BBP/users/f:9d46ddd6-134e-44d6-aa74-bdf00f48dfce:sy',
      _updatedAt: '2018-11-15T08:40:52.735Z',
      _updatedBy:
        'https://bbp.epfl.ch/nexus/v1/realms/BBP/users/f:9d46ddd6-134e-44d6-aa74-bdf00f48dfce:sy',
      _rev: 2,
      _deprecated: false,
    },
  ],
  _total: 1,
};

// [KP] The mappings property here is verbose, but it's exactly what we expect
// from a default ESView response, so I think it's useful to keep here as-is.
export const mockViewsListResponse: ViewsListResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/search.json',
    'https://bluebrain.github.io/nexus/contexts/resource.json',
    'https://bluebrain.github.io/nexus/contexts/view.json',
  ],
  _total: 2,
  _results: [
    {
      '@id': 'nxv:defaultElasticSearchIndex',
      '@type': ['View', 'Alpha', 'ElasticSearchView'],
      _uuid: '684bd815-9273-46f4-ac1c-0383d4a98254',
      includeMetadata: true,
      mapping: {
        dynamic: false,
        properties: {
          '@id': {
            type: 'keyword',
          },
          '@type': {
            type: 'keyword',
          },
          _distribution: {
            properties: {
              '@id': {
                type: 'keyword',
              },
              '@type': {
                type: 'keyword',
              },
              _byteSize: {
                type: 'long',
              },
              _digest: {
                properties: {
                  _algorithm: {
                    type: 'keyword',
                  },
                  _value: {
                    type: 'keyword',
                  },
                },
                type: 'nested',
              },
              _downloadURL: {
                type: 'keyword',
              },
              _mediaType: {
                type: 'keyword',
              },
              _originalFileName: {
                type: 'keyword',
              },
            },
            type: 'nested',
          },
          _original_source: {
            type: 'text',
          },
          _self: {
            type: 'keyword',
          },
          _constrainedBy: {
            type: 'keyword',
          },
          _project: {
            type: 'keyword',
          },
          _createdAt: {
            type: 'date',
          },
          _createdBy: {
            type: 'keyword',
          },
          _updatedAt: {
            type: 'date',
          },
          _updatedBy: {
            type: 'keyword',
          },
          _rev: {
            type: 'long',
          },
          _deprecated: {
            type: 'boolean',
          },
        },
      },
      sourceAsText: true,
      _rev: 1,
      _deprecated: false,
    },
    {
      '@id': 'nxv:defaultSparqlIndex',
      '@type': ['View', 'SparqlView'],
      _uuid: 'd88b71d2-b8a4-4744-bf22-2d99ef5bd26b',
      _rev: 1,
      _deprecated: false,
    },
  ],
};

export const mockElasticSearchViewResponse: ElasticSearchViewResponse = mockViewsListResponse
  ._results[0] as ElasticSearchViewResponse;

export const mockSparqlViewResponse: SparqlViewResponse = mockViewsListResponse
  ._results[1] as SparqlViewResponse;

export const mockElasticSearchViewAggregationResponse: ElasticSearchViewAggregationResponse = {
  aggregations: {
    schemas: {
      buckets: [
        {
          doc_count: 971,
          key: 'https://neuroshapes.org/dash/reconstructedpatchedcell',
        },
        {
          doc_count: 701,
          key: 'https://neuroshapes.org/dash/labeledcell',
        },
        {
          doc_count: 661,
          key: 'https://neuroshapes.org/dash/patchedcell',
        },
        {
          doc_count: 401,
          key: 'https://neuroshapes.org/dash/slice',
        },
        {
          doc_count: 396,
          key: 'https://neuroshapes.org/dash/subject',
        },
        {
          doc_count: 119,
          key: 'https://neuroshapes.org/dash/reconstruction',
        },
        {
          doc_count: 74,
          key: 'https://bluebrain.github.io/nexus/schemas/shacl-20170720.ttl',
        },
        {
          doc_count: 15,
          key: 'https://neuroshapes.org/dash/person',
        },
        {
          doc_count: 3,
          key: 'https://bluebrain.github.io/nexus/schemas/resource.json',
        },
      ],
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
    },
  },
};

export const mockElasticSearchViewQueryResponse: ElasticSearchViewQueryResponse = {
  _shards: {
    failed: 0,
    skipped: 0,
    successful: 0,
    total: 0,
  },
  hits: {
    hits: [
      {
        _score: 1,
        _id:
          'https://bbp.epfl.ch/nexus/v0/data/bbp/experiment/subject/v0.1.0/eb1b8cbb-ae00-44c0-889e-107e9f16faef',
        _index:
          'kg_v1_ed2a98da-e96f-4a7c-9362-f8e07a9e9151_684bd815-9273-46f4-ac1c-0383d4a98254_1',
        _source: {
          '@id':
            'https://bbp.epfl.ch/nexus/v0/data/bbp/experiment/subject/v0.1.0/eb1b8cbb-ae00-44c0-889e-107e9f16faef',
          '@type': [
            'http://www.w3.org/ns/prov#Entity',
            'https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/Subject',
          ],
          _original_source:
            '{"@context":{"xml":"http://www.w3.org/XML/1998/namespace","rdf":"http://www.w3.org/1999/02/22-rdf-syntax-ns#","rdfs":"http://www.w3.org/2000/01/rdf-schema#","xsd":"http://www.w3.org/2001/XMLSchema#","owl":"http://www.w3.org/2002/07/owl#","skos":"http://www.w3.org/2004/02/skos/core#","prov":"http://www.w3.org/ns/prov#","dcat":"http://www.w3.org/ns/dcat#","sh":"http://www.w3.org/ns/shacl#","shsh":"http://www.w3.org/ns/shacl-shacl#","dcterms":"http://purl.org/dc/terms/","schema":"http://schema.org/","nxv":"https://bbp-nexus.epfl.ch/vocabs/nexus/core/terms/v0.1.0/","nsg":"https://bbp-nexus.epfl.ch/vocabs/bbp/neurosciencegraph/core/v0.1.0/","this":"http://example.org/minds/","dc":"http://purl.org/dc/elements/1.1/","vann":"http://purl.org/vocab/vann/","void":"http://rdfs.org/ns/void#","Dataset":{"@id":"dcat:Dataset"},"valueZ":{"@id":"nsg:valueZ"},"digest":{"@id":"nxv:digest"},"maxValue":{"@id":"schema:maxValue"},"unitCode":{"@id":"schema:unitCode"},"contribution":{"@id":"nsg:contribution"},"type":{"@id":"rdf:type"},"license":{"@id":"dcat:license"},"accessURL":{"@id":"dcat:accessURL"},"subject":{"@id":"nsg:subject"},"label":{"@id":"rdfs:label"},"valueX":{"@id":"nsg:valueX"},"species":{"@id":"nsg:species"},"valueY":{"@id":"nsg:valueY"},"repository":{"@id":"nsg:repository"},"agent":{"@id":"prov:agent"},"objectOfStudy":{"@id":"nsg:objectOfStudy"},"byteSize":{"@id":"dcat:byteSize"},"coordinatesInBrainAtlas":{"@id":"nsg:coordinatesInBrainAtlas"},"value":{"@id":"schema:value"},"downloadURL":{"@id":"dcat:downloadURL"},"atlasSpatialReferenceSystem":{"@id":"nsg:atlasSpatialReferenceSystem"},"mediaType":{"@id":"dcat:mediaType"},"brainLocation":{"@id":"nsg:brainLocation"},"distribution":{"@id":"dcat:distribution"},"brainRegion":{"@id":"nsg:brainRegion"},"minValue":{"@id":"schema:minValue"}},"@id":"https://bbp.epfl.ch/nexus/v0/data/bbp/experiment/subject/v0.1.0/eb1b8cbb-ae00-44c0-889e-107e9f16faef","@type":["nsg:Subject","prov:Entity"],"schema:name":"Rattus norvegicus Wistar Han","nsg:providerId":"C300797C","species":{"@id":"http://purl.obolibrary.org/obo/NCBITaxon_10116","label":"Rattus norvegicus"},"nsg:strain":{"@id":"http://purl.obolibrary.org/obo/RS_0001833","label":"Wistar Han"}}',
          _self:
            'https://bbp.epfl.ch/nexus/v1/resources/anorg/testcore/datashapes:subject/https%3A%2F%2Fbbp.epfl.ch%2Fnexus%2Fv0%2Fdata%2Fbbp%2Fexperiment%2Fsubject%2Fv0.1.0%2Feb1b8cbb-ae00-44c0-889e-107e9f16faef',
          _constrainedBy: 'https://neuroshapes.org/dash/subject',
          _project: 'https://bbp.epfl.ch/nexus/v1/projects/anorg/testcore',
          _createdAt: '2018-11-21T00:02:27.206Z',
          _createdBy:
            'https://bbp.epfl.ch/nexus/v1/realms/BBP/users/f:9d46ddd6-134e-44d6-aa74-bdf00f48dfce:sy',
          _updatedAt: '2018-11-26T11:00:29.200Z',
          _updatedBy:
            'https://bbp.epfl.ch/nexus/v1/realms/BBP/users/f:9d46ddd6-134e-44d6-aa74-bdf00f48dfce:sy',
          _rev: 11,
          _deprecated: false,
        },
        _type: 'doc',
      },
    ],
    max_score: 1,
    total: 3341,
  },
  timed_out: false,
  took: 8,
};
