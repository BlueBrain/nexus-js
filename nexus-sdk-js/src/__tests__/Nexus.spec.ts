import { resetMocks, mock, mockResponse, mockResponses } from 'jest-fetch-mock';
import Nexus from '..';
import { mockOrgResponse, mockListProjectResponse } from '../__mocks__/helpers';
import Organization from '../Organization';

const nexus = new Nexus({ environment: 'http://example.com' });
describe('Nexus class', () => {
  it('should create a Nexus instance', () => {
    expect(nexus).toBeInstanceOf(Nexus);
  });
  it('should throw an error', () => {
    // @ts-ignore
    expect(() => new Nexus()).toThrowErrorMatchingSnapshot();
    // @ts-ignore
    expect(() => new Nexus({ foo: 'bar' })).toThrowErrorMatchingSnapshot();
  });

  describe('get an Org', () => {
    afterEach(() => {
      resetMocks();
    });
    it('should call the API', () => {
      nexus.getOrganization('myorg');
      expect(mock.calls.length).toBe(1);
    });
    it('should return an organisation', async () => {
      mockResponse(JSON.stringify(mockOrgResponse));
      const org: Organization = await nexus.getOrganization('myorg');
      expect(org).toBeInstanceOf(Organization);
      expect(org.label).toEqual('myorg');
    });
  });

  describe('list a list of Org', () => {
    afterEach(() => {
      resetMocks();
    });
    it('should return a list of organisations', async () => {
      mockResponses(
        [JSON.stringify(mockListProjectResponse)],
        [JSON.stringify(mockOrgResponse)],
        [JSON.stringify(mockOrgResponse)],
      );
      const orgs: Organization[] = await nexus.listOrganizations();
      expect(mock.calls.length).toBe(3);
      expect(orgs.length).toEqual(1);
    });
  });
});
