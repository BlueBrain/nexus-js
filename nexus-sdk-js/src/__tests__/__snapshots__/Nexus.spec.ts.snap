// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nexus class should throw an error 1`] = `"You need to provide a Nexus config."`;

exports[`Nexus class should throw an error 2`] = `"No environment provided. Please specify your Nexus instance endpoint."`;
